scripts/train.ipynb

```diff
@@ -312,48 +775,82 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "def graph_to_composite_arrow(graph: DirectedGraph):\n",
-    "    nodes = [(src, dst) for src, dst, _ in graph.edges(data=True)]\n",
-    "    edges = [list(rel.keys()) for _, _, rel in graph.edges(data=True)]\n",
-    "    return nodes, edges"
+    "def epoque_valid():\n",
+    "    # for (5174 / n_iter) * 2 = ~320 batches (best case)\n",
+    "    batches = 10\n",
+    "    print_each_nth = 5\n",
+    "    range_with_progress = trange(batches)\n",
+    "    for i in trange(batches, desc='Batch valid'):\n",
+    "        sampled_subgraph = random_walk_vertex_sample(graph=graph_valid, rng=random.Random(), n_iter=32, n_seeds=1,  use_opposite=True)\n",
+    "        plot_subgraphs(sampled_subgraph, batch_n=i, plot_nth=print_each_nth, graph_info=False)\n",
+    "        nodes, edges = graph_to_nodes_edges(sampled_subgraph)\n",
+    "        arrows = nodes_edges_to_arrows(nodes, edges)\n",
+    "        cache, matches = model.train(\n",
+    "            data_points = ds.entity_id2vec,\n",
+    "            relations = arrows,\n",
+    "            # NOTE: Labels could be a couplete graph, a subgraph from random_walk or a sub-sub-graph used to create a batch\n",
+    "            labels = sampled_subgraph,\n",
+    "            step = False,\n",
+    "            match_negatives=False\n",
+    "        )\n",
+    "        log_results(cache, matches, epoque='valid')"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
    "metadata": {
+    "id": "IOPyRB1KYmGf",
     "tags": []
    },
    "outputs": [],
    "source": [
     "# NOTE: for large graphs, random_walk functions family can be used to sub-sample graph\n",
     "# while preserving its topology \n",
-    "for i in tnrange(10, desc='1st loop: batches'):\n",
-    "    arrows_graph = uniform_sample(graph=graph_labels, sample_vertices_size=7, rng=random.Random(), with_edges=True)\n",
-    "    nodes, edges = graph_to_composite_arrow(arrows_graph)\n",
-    "    if edges:\n",
-    "        arrows = [CompositeArrow(nodes=node, arrows=edge) for node, edge in zip(nodes, edges)]\n",
-    "    else:\n",
-    "        arrows = [CompositeArrow(nodes=node) for node in nodes]\n",
-    "    cache, matches = model.train(\n",
-    "        data_points = ds.entity_id2vec,\n",
-    "        relations = arrows,\n",
-    "        # NOTE: Labels could be a couplete graph, a subgraph from random_walk or a sub-sub-graph used to create a batch\n",
-    "        labels = arrows_graph,\n",
-    "        step = True,\n",
-    "        match_negatives=False\n",
-    "    )\n",
-    "    log_results(cache, matches)"
+    "epoques = 1\n",
+    "for j in trange(epoques, desc='Epoque'):\n",
+    "    epoque_train()\n",
```

Comment:
I get an error from wandb_logger when using it (basically related to the fact that Parameters objects are not serializable).
I had this problem when I created the logger but used a cast to Tensor class to fix it. You should add that cast again.